File: Coding Competitions/ACM - International Collegiate Programming Contest 2014 - Pacific NW Region Programming Contest/L.py
#team203
import math
def distance(star1,star2):
	result = math.sqrt( (star1[0] - star2[0])**2 +  
			(star1[1] - star2[1])**2 +
			(star1[2] - star2[2])**2 )
	if result - int(result) > 0.5:
		result = int(result) + 1
	else:
		result = int(result)
	return result


T = input()
kase = 1
while T:
	T -= 1
	NAME = {}
	POS = []
	star = input()
	for i in range(0,star):
		Line = raw_input().split()
		pos = map(int,Line[1:])
		NAME[Line[0]] = i
		POS.append(pos)
	dis = []
	for i in range(0,star):
		row = []
		for j in range(0,star):
			row.append(0.0)
		dis.append(row)

	for i in range(0,star):
		for j in range(0,star):	
			if i == j:
				dis[i][j] = 0.0
			else:
				dis[i][j] = dis[j][i] = distance(POS[i],POS[j])
	#	
	#for i in range(0,star):
	#	for j in range(0,star):	
	#		print dis[i][j],
	#	print ""
	#
	#

	worm = input()
	for i in range(0,worm):
		F,S = raw_input().split()
		dis[NAME[F]][NAME[S]] = 0.0
	for i in range(0,star):
		for j in range(0,star):
			for k in range(0,star):
				dis[i][j] = min(dis[i][j],(dis[i][k] + dis[k][j]))
	
	print "Case %d:" % kase
	kase += 1
	query = input()
	for i in range(0,query):
		F,S = raw_input().split()
		result = dis[NAME[F]][NAME[S]]
		
		print "The distance from "+ F + " to " + S + " is",
		print int(result),
		print "parsecs."

File: Coding Competitions/ACM - International Collegiate Programming Contest 2014 - Pacific NW Region Programming Contest/L.cpp
#include <bits/stdc++.h>
using namespace std;

struct star{
	long long x,y,z;
}Star[100];

long long Distance(struct star &star1, struct star &star2){
	double x = sqrt( (star1.x - star2.x)*(star1.x - star2.x) +  
			(star1.y - star2.y)*(star1.y - star2.y) +
			(star1.z - star2.z)*(star1.z - star2.z) );
	if( (x - (long long) x) > 0.5 )
				x = x + 1; 
	return x;
}

int main(){
	//ios_base::sync_with_stdio(0);
	int T;
	int kase = 1;
	cin >> T;
	while(T--){
		int n;
		long long x,y,z;
		double dis[100][100] = {0.0};
		map<string, int> NAME;
		cin >> n;	
		string name;
		for(int i = 0; i < n; i++){
			cin >> name >> x >> y >> z;
			NAME[name] = i;
			Star[i].x = x;
			Star[i].y = y;
			Star[i].z = z;
		}
		for(int i = 0; i < n; i++){
			for(int j = 0; j < n; j++){
				if (i == j){
					dis[i][j] = 0;
				}
				else{
					dis[i][j] = dis[j][i] = Distance(Star[i],Star[j]);
				}
			}
		}
		
		int worm;
		cin >> worm;
		string F;
		string S;
		for(int i = 0; i < worm; i++){
			cin >> F >> S;
			dis[NAME[F]][NAME[S]] = 0.0;
		}
		
		for(int i = 0; i < n; i++){
			for(int j = 0; j < n; j++){
				for(int k = 0; k < n; k++){
					dis[i][j] = min(dis[i][j], dis[i][k] + dis[k][j]);
				}
			}
		}
		
		/*
		for(int i = 0; i < n; i++){
			for(int j = 0; j < n; j++){
				cout << dis[i][j] << " " ;
			}
			cout << endl;
		}
		*/

		cout << "Case " << kase++ <<":" << endl;
		int query;
		string queryf;
		string querys;

		cin >> query;
		for(int i = 0; i< query; i++){
			long long RESULT;
			RESULT = dis[NAME[queryf]][NAME[querys]];
			/*
			if( (RESULT - (long long) RESULT) > 0.5 )
				RESULT = RESULT + 1;
				*/
			cin >> queryf >> querys;
			cout << "The distance from " << queryf << " to " << querys << " is " 
			<< RESULT << " parsecs." << endl;
		}	
		
	}
	return 0;
}
